File: backend/internal/service/display_currency.go
package service

import (
	"strconv"
	"strings"
)

// Display currency is what a panel *shows*, and it is deliberately not the same
// concept as the payment currency a gateway *collects*.
//
// Every amount in this system is stored in USD (see `payment_orders.amount`,
// `users.balance`, `vip_tiers.min_spend_usd`). A display currency never changes
// a stored number; it only decides how that number is rendered for the reader,
// and which unit the recharge screen quotes before the gateway takes over.
//
// The mapping is by UI locale rather than by account or IP, because the locale
// is the one signal the reader controls directly: a Vietnamese reader who
// switches to English is asking to see dollars, and honouring that is less
// surprising than pinning the currency to a profile field they cannot see.
const (
	// DisplayCurrencyUSD is the canonical unit — the rate is always exactly 1,
	// so a missing or broken FX feed degrades to "show the stored number".
	DisplayCurrencyUSD = "USD"
	DisplayCurrencyCNY = "CNY"
	DisplayCurrencyVND = "VND"
)

// SettingDisplayUSDToCNYRate holds units of CNY per 1 USD, for the `zh` locale's
// display currency.
//
// It has no VND-style twin key because VND display reuses
// SettingSubscriptionUSDToVNDRate on purpose: a Vietnamese reader must see the
// same number on the plan card that the SePay gateway will actually charge, and
// two independently-maintained rates guarantee those two numbers drift apart.
const SettingDisplayUSDToCNYRate = "DISPLAY_USD_TO_CNY_RATE"

// DisplayFXRates maps a display currency to how many of its units one USD buys.
// USD is always present at 1. A currency is omitted entirely when no usable rate
// is known — the frontend then falls back to rendering USD rather than inventing
// a conversion, so a stale or unconfigured feed can never quote a wrong price.
type DisplayFXRates map[string]float64

// BuildDisplayFXRates assembles the public rate table from the two stored rate
// settings. Non-positive rates are treated as "not configured", which is the
// same convention normalizeUSDToVNDRate already uses for the checkout path.
func BuildDisplayFXRates(usdToCNY, usdToVND float64) DisplayFXRates {
	rates := DisplayFXRates{DisplayCurrencyUSD: 1}
	if usdToCNY > 0 {
		rates[DisplayCurrencyCNY] = usdToCNY
	}
	if usdToVND > 0 {
		rates[DisplayCurrencyVND] = usdToVND
	}
	return rates
}

// parseDisplayRate reads a stored rate setting. Anything unparseable or
// non-positive collapses to 0, i.e. "no rate", rather than to an error: a
// corrupt row must not take down the public settings endpoint.
func parseDisplayRate(raw string) float64 {
	value, err := strconv.ParseFloat(strings.TrimSpace(raw), 64)
	if err != nil || value <= 0 {
		return 0
	}
	return value
}

// crossRateFromVND converts two "VND per unit" board quotes into a direct
// USD→CNY rate. The bank publishes every currency against the dong, so the dong
// cancels: (VND per USD) / (VND per CNY) = CNY per USD.
//
// Returning 0 rather than an error keeps a missing CNY row from failing the VND
// sync that shares the same fetch.
func crossRateFromVND(vndPerUSD, vndPerCNY float64) float64 {
	if vndPerUSD <= 0 || vndPerCNY <= 0 {
		return 0
	}
	return vndPerUSD / vndPerCNY
}

File: frontend/src/utils/displayCurrency.ts
/**
 * Display currency — what a price is *shown* in, which is deliberately not the
 * same thing as the currency a payment gateway *collects*.
 *
 * Every amount this panel holds is USD: balances, plan prices, VIP spend
 * thresholds, order amounts. Converting is a rendering concern only — nothing
 * here ever feeds a stored value or a charge. The gateway conversion lives in
 * `views/user/PaymentView.vue` (mirroring the backend's
 * `calculateGatewayBaseAmount`) and is driven by the *provider's* currency, not
 * by this module.
 *
 * The currency follows the UI locale rather than the account or the IP, because
 * the locale is the one signal the reader controls directly: someone who
 * switches to English is asking for dollars, and honouring that is less
 * surprising than pinning currency to a profile field they cannot see.
 *
 * Rates arrive from `GET /settings/public` (`display_fx_rates`) as "units per 1
 * USD". A currency with no usable rate is simply absent from that table, and
 * every function here then falls back to USD — a price that is honestly in
 * dollars beats one that is silently wrong in dong.
 */
import { formatPaymentAmount, paymentCurrencyFractionDigits } from '@/components/payment/currency'
import { DEFAULT_LOCALE, type LocaleCode } from '@/i18n/locale'

export type DisplayCurrency = 'USD' | 'CNY' | 'VND'

/** The canonical unit: every stored amount is already in it, so its rate is 1. */
export const BASE_DISPLAY_CURRENCY: DisplayCurrency = 'USD'

export const LOCALE_DISPLAY_CURRENCY: Record<LocaleCode, DisplayCurrency> = {
  en: 'USD',
  zh: 'CNY',
  vi: 'VND',
}

/** Units of each display currency per 1 USD. Mirrors the backend payload. */
export type DisplayFXRates = Record<string, number>

/** BCP-47 tag for `Intl`, so grouping and the symbol match the rendered locale. */
const INTL_LOCALE: Record<LocaleCode, string> = {
  en: 'en-US',
  zh: 'zh-CN',
  vi: 'vi-VN',
}

/**
 * Takes a plain `string` rather than a `LocaleCode` so callers holding
 * `useI18n().locale` (typed `string`) can use it without narrowing at every
 * call site — the alternative is a ternary repeated next to every `Intl` call,
 * which is exactly how `'zh-CN' : 'en-US'` ended up hardcoded in two views.
 */
export function intlLocaleFor(locale: string): string {
  return INTL_LOCALE[locale as LocaleCode] ?? INTL_LOCALE[DEFAULT_LOCALE]
}

/** The currency a locale *wants*, before checking whether a rate exists for it. */
export function preferredDisplayCurrency(locale: LocaleCode): DisplayCurrency {
  return LOCALE_DISPLAY_CURRENCY[locale] ?? BASE_DISPLAY_CURRENCY
}

function usableRate(rates: DisplayFXRates | null | undefined, currency: string): number {
  const rate = rates?.[currency]
  return typeof rate === 'number' && Number.isFinite(rate) && rate > 0 ? rate : 0
}

/**
 * What this locale can actually be priced in right now.
 *
 * Resolution, not preference: an operator who never enabled the FX sync has no
 * VND rate stored, and a Vietnamese reader must then see dollars rather than a
 * number derived from a rate of zero.
 */
export function resolveDisplayCurrency(
  locale: LocaleCode,
  rates: DisplayFXRates | null | undefined,
): { currency: DisplayCurrency; rate: number; converted: boolean } {
  const preferred = preferredDisplayCurrency(locale)
  if (preferred === BASE_DISPLAY_CURRENCY) {
    return { currency: BASE_DISPLAY_CURRENCY, rate: 1, converted: false }
  }

  const rate = usableRate(rates, preferred)
  if (rate <= 0) {
    return { currency: BASE_DISPLAY_CURRENCY, rate: 1, converted: false }
  }

  return { currency: preferred, rate, converted: true }
}

/**
 * Convert a stored USD amount into the display currency, rounded to that
 * currency's minor unit.
 *
 * Rounding here rather than at render time is what keeps a total from
 * disagreeing with the sum of the rows it is made of: two unrounded dong
 * figures printed as whole numbers can add up to one more dong than their
 * printed total.
 */
export function convertFromUSD(amountUSD: number, rate: number, currency: string): number {
  if (!Number.isFinite(amountUSD)) {
    return 0
  }
  if (!Number.isFinite(rate) || rate <= 0) {
    return amountUSD
  }

  const digits = paymentCurrencyFractionDigits(currency)
  const factor = 10 ** digits
  return Math.round(amountUSD * rate * factor) / factor
}

/**
 * Render a stored USD amount in the locale's display currency.
 *
 * Falls back to a plain USD rendering whenever no rate is available, so a
 * missing feed degrades to "correct but in dollars" rather than to a wrong
 * number or an empty cell.
 */
export function formatDisplayAmount(
  amountUSD: number,
  locale: LocaleCode,
  rates: DisplayFXRates | null | undefined,
): string {
  const { currency, rate } = resolveDisplayCurrency(locale, rates)
  return formatPaymentAmount(convertFromUSD(amountUSD, rate, currency), currency, intlLocaleFor(locale))
}
